From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Robert Crabtree <robertx.l.crabtree@intel.com>
Date: Wed, 25 Jul 2012 17:53:12 -0700
Subject: [PATCH 01/15] SW encode: Fix output buffer size

The output buffer size needs to be updated if the resolution is updated.
The output buffer was configured for QCIF resolution when the actual
resolution was VGA. Buffer overflow was resulting.

Change-Id: Ieca3943d8d67bfcfcf49fed95d17be712f4b14b5
Signed-off-by: Robert Crabtree <robertx.l.crabtree@intel.com>
---
 .../omx/SoftVideoEncoderOMXComponent.cpp      | 27 +++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/media/libstagefright/omx/SoftVideoEncoderOMXComponent.cpp b/media/libstagefright/omx/SoftVideoEncoderOMXComponent.cpp
index d75acdabf8..11e563cdc2 100644
--- a/media/libstagefright/omx/SoftVideoEncoderOMXComponent.cpp
+++ b/media/libstagefright/omx/SoftVideoEncoderOMXComponent.cpp
@@ -377,6 +377,33 @@ OMX_ERRORTYPE SoftVideoEncoderOMXComponent::internalGetParameter(
             return OMX_ErrorNone;
         }
 
+        case OMX_IndexParamPortDefinition:
+        {
+            OMX_PARAM_PORTDEFINITIONTYPE *def =
+                (OMX_PARAM_PORTDEFINITIONTYPE *)param;
+
+            if (def->nPortIndex > 1) {
+                return OMX_ErrorUndefined;
+            }
+
+            OMX_ERRORTYPE err = SimpleSoftOMXComponent::internalGetParameter(index, param);
+            if (OMX_ErrorNone != err) {
+                return err;
+            }
+
+            def->format.video.nFrameWidth = mWidth;
+            def->format.video.nFrameHeight = mHeight;
+
+            // XXX: For now just configure input and output buffers the same size.
+            // May want to determine a more suitable output buffer size independent
+            // of YUV format.
+            CHECK(mColorFormat == OMX_COLOR_FormatYUV420Planar ||
+                    mColorFormat == OMX_COLOR_FormatYUV420SemiPlanar);
+            def->nBufferSize = mWidth * mHeight * 3 / 2;
+
+            return OMX_ErrorNone;
+        }
+
         default:
             return SimpleSoftOMXComponent::internalGetParameter(index, param);
     }
